games/checkers/piece.py

```diff
@@ -0,0 +1,183 @@
+from enum import Enum, auto
+from abc import ABC, abstractmethod
+from typing import Iterable, Protocol, Sequence, Tuple, Type, TypeVar
+from games.checkers import EMPTY_SYMBOL
+import numpy as np
+
+
+class PlayerColor(Enum):
+    WHITE = auto()
+    BLACK = auto()
+
+
+TPiece = TypeVar('TPiece', bound='CheckersPiece')
+
+
+class CheckersPiece(ABC):
+
+    def __init__(self, row: int, col: int):
+        self.row = row
+        self.col = col
+
+    @property
+    @abstractmethod
+    def color(self) -> PlayerColor:
+        ...
+
+    @property
+    @abstractmethod
+    def symbol(self) -> str:
+        ...
+
+    def move_to(self: TPiece, board: np.ndarray, row, col) -> Tuple[np.ndarray, 'TPiece']:
+        board = board.copy()
+        board[self.row, self.col] = EMPTY_SYMBOL
+        if self._is_jump(board, row, col):
+            self._jump(board, row, col)
+        else:
+            self._move(board, row, col)
+        return board, board[row, col]
+
+    def _is_jump(self, board: np.ndarray, new_row, new_col) -> bool:
+        return manhattan_distance((new_row, new_col), (self.row, self.col)) != 2
+
+    def _jump(self, board: np.ndarray, new_row, new_col):
+        r, c = self._get_jump_direction(new_row, new_col)
+        beaten_cell = self.row + r, self.col + c
+        board[beaten_cell] = EMPTY_SYMBOL
+        self._move(board, new_row, new_col)
+
+    def _get_jump_direction(self, new_row: int, new_col: int):
+        delta_row, delta_col = new_row - self.row, new_col - self.col
+        delta_row = np.sign(delta_row) * (delta_row // delta_row) if delta_row != 0 else 0
+        delta_col = np.sign(delta_col) * (delta_col // delta_col) if delta_col != 0 else 0
+        return delta_row, delta_col
```

Comment:
Ta funkcja wylicza wektor jednostkowy od `self.row, self.col` do `new_row, new_col`
To jest przydatne, żeby wiedzieć przez którą kratkę odbywa się skok.